### PR TITLE
Oefimenko

### PR DESCRIPTION
**Pull request by oiefimenko**
Hometask 4

### CODE REVIEW

**Comment by nonkor** [comment]
@oefimenko Wow! I like your style so much! Amazing work, really.
But there are few issues which I have to note.

File: lesson4/hometask.rb [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    self.__setobj__(initial_hash)
+  end
+
+  def []=(key, value)
+    raise ArgumentError, 'Unable to symbolize key!' unless key.respond_to?(:to_sym)
+    super(key.to_sym, value)
+    unless self.__getobj__.respond_to?(key.to_sym)
+      define_singleton_method(key.to_sym) do
+        obj = self.__getobj__[key]
+        obj.is_a?(Proc) ? self.__getobj__[key] = obj.call : obj
```
Comment (nonkor) [inline lesson4/hometask.rb]: There is hidden issue here.
Initially you call `super(key.to_sym, value)` that means you store `key` as symbol.
But.
In next lines you check value type. If it's a `proc` you define method to call that proc. It's ok, but how do you do this? Take a look on `self.__getobj__[key]`. Key could be a string here. And we could store value after block performing associated with string key. So, we DO NOT rewrite it, we create a new pair in that case.

Try to run this:

``` ruby
lh = LazyHashDelegator.new({})
lh['test'] = proc { puts 'test output' }
p lh # => {:test=>#<Proc:0x007fd5ac04cf48@temp.rb:80>}
lh.test # => test output
p lh # => {:test=>#<Proc:0x007fd5ac04cf48@temp.rb:80>, "test"=>nil}
```

File: lesson4/hometask.rb [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+#==============================================
+# SimpleDelegator usage and no method missing
+#==============================================
+
+require 'delegate'
+
+class LazyHashDelegator < SimpleDelegator
+  def initialize(initial_hash = {})
+    self.__setobj__(initial_hash)
```
Comment (nonkor) [inline lesson4/hometask.rb]: initializer could be redundant if your test allows as to specify `hash` as decorated object directly, without using a default value)

File: lesson4/hometask.rb [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  end
+
+  def []=(key, value)
+    raise ArgumentError, 'Unable to symbolize key!' unless key.respond_to?(:to_sym)
+    super(key.to_sym, value)
+    unless self.__getobj__.respond_to?(key.to_sym)
+      define_singleton_method(key.to_sym) do
+        obj = self.__getobj__[key]
```
Comment (nonkor) [inline lesson4/hometask.rb]: why not just `self[key]`? we delegated that object already

File: lesson4/hometask.rb [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    end
+  end
+
+  def [](key)
+    raise ArgumentError, 'Unable to symbolize key!' unless key.respond_to?(:to_sym)
+    super(key.to_sym)
+  end
+
+  def lazy?(key)
+    self.__getobj__[key].is_a?(Proc)
```
Comment (nonkor) [inline lesson4/hometask.rb]: the same here, see [remark above](https://github.com/nonkor/Betlab-ruby-courses/pull/16/files#r78758432) ^
I believe we should not use technical methods in cases where we can avoid this.

File: lesson4/hometask.rb [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+  result = result && lazy_hash[:proc] == lazy_hash.proc
+  result = result && lazy_hash[:lambda] == lazy_hash.lambda
+  result = result && lazy_hash[:raw] == lazy_hash.raw
+
+  result = result && !lazy_hash.lazy?(:proc)
+  result = result && !lazy_hash.lazy?(:lambda)
+  result = result && !lazy_hash.lazy?(:raw)
+
+  puts result
```
Comment (nonkor) [inline lesson4/hometask.rb]: So delicious way to check all `true` expectations! I need to remember it)

File: lesson4/hometask.rb [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+class LazyHashDelegator < SimpleDelegator
+  def initialize(initial_hash = {})
+    self.__setobj__(initial_hash)
+  end
+
+  def []=(key, value)
+    raise ArgumentError, 'Unable to symbolize key!' unless key.respond_to?(:to_sym)
+    super(key.to_sym, value)
+    unless self.__getobj__.respond_to?(key.to_sym)
```
Comment (nonkor) [inline lesson4/hometask.rb]: again redundant `__getobj__`

File: lesson4/hometask.rb [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+      super
+    end
+  end
+
+  def lazy?(key)
+    self[key].is_a?(Proc)
+  end
+end
+
+test(LazyHash)
```
Comment (nonkor) [inline lesson4/hometask.rb]: use smth like `"ensure_newline_at_eof_on_save": true` in your editor

**Comment by oiefimenko** [comment]
Fixed eof, key symbolizing and redundant calls to **getobj**
As for overriding constructor in LazyHashDelegator - it's allowing to use it as Decorator like
`
lazy_hash = LazyHashDelegator.new(Hash.new)
`
or as Hash inherited object itself, like
`
lazy_hash = LazyHashDelegator.new
`
So I was looking for some flexibility

**Comment by nonkor** [comment]
@oefimenko I see your point about initializer 👍 

But I realized one more thing - there is an issue in a case when we set up not empty hash initially.
Like `lh = LazyHashDelegator.new(a: proc {return 1}, b: 2)`
We cannot call `lh.a` because related method is still not defined. Makes sense to update `#initialize` to take it into consideration.

**Pull request merged**